### PR TITLE
Add or create ImagePullSecret  in a run collector 

### PR DESCRIPTION
**Pull request by manavellamnimble**
@marccampbell @dexhorthy regarding issue #247   I added the possibility to either add an existing secret :

```yaml
- run
     ...
     imagePullSecret:
         name: mysecret
```
or to create a secret. In this case there are various options:
- add an opaque secret, with one or many data fields in it
- optionally add a name to the secret. If none is provided a random name is generated 
- provide a type kubernetes.io/dockerconfigjson secret, in which case there are two options
   - provide the path to the config file. The client parses and encodes it automatically.
   - provide an already base64-encoded configfile. Though it is not actually an option when using the k8s client, I dealt with this because it is the suggested way in kubernetes official docs (but it only works when using yaml). What I did was to decode the string, indent it to rebuild the json, and put it into the secret. Base64 does not takes into consideration white spaces or tabs, making this algorithm independent of  the original config file's indentation. All this thinking in making it more familiar to the user.

I handle the secret's data as []byte because it is safer according to what I read.
```yaml
- run
      ...
      imagePullSecret:
          name: mysecret
          data: 
            .dockerconfigjson: path/.docker/config.json
          type: kubernetes.io/dockerconfigjson
```
```yaml
- run
      ...
      imagePullSecret:
          name: mysecret
          data: 
            .dockerconfigjson: woJICJhdXRocyI6IHsKCQkiaHR0cHM6Ly9pbmRleC5kb2NrZXIuaW8vdjEvIjoge30KCX0sCgkiSHR0cEhlYWRlcnMiOiB7CgkJIlVzZXItQWdlbnQiOiAiRG9ja2VyLUNsaWVudC8xOS4wMy4xMiAoZGFyd2luKSIKCX0sCgkiY3JlZHNTdG9yZSI6ICJkZXNrdG9wIiwKCSJleHBlcmltZW50YWwiOiAiZGlzYWJsZWQiLAoJInN0YWNrT3JjaGVzdHJhdG9yIjogInN3YXJtIgp9
          type: kubernetes.io/dockerconfigjson
```
                         

```yaml
- run
      ...
      imagePullSecret:
          data: 
            a: foo
            b: bar
```
Let me know what you think of the solution!

### CODE REVIEW

File: pkg/collect/run.go [hunk 1]
```diff
@@ -137,10 +149,79 @@ func runPod(ctx context.Context, client *kubernetes.Clientset, runCollector *tro
 		},
 	}
 
+	if runCollector.ImagePullSecret != nil {
+		err := createSecret(ctx, client, runCollector.ImagePullSecret, &pod)
+		if err != nil {
+			return nil, err
```
Comment (divolgin) [inline pkg/collect/run.go]: needs `errors.Wrap`

File: pkg/collect/run.go [hunk 2]
```diff
@@ -137,10 +149,79 @@ func runPod(ctx context.Context, client *kubernetes.Clientset, runCollector *tro
...
+			if err != nil {
+				//If data is not a valid path, we assume data is a base64 encoded config.json file
+				//Client only accepts Json formated files as data, so we decode and indent it (indentation is required)
+				parsedConfig, err := base64.StdEncoding.DecodeString(v)
+				if err != nil {
+					return err
+				}
+				err = json.Indent(&out, parsedConfig, "", "\t")
+				if err != nil {
+					return errors.Errorf("Secret's config file not found or unable to parse encoded data.")
```
Comment (divolgin) [inline pkg/collect/run.go]: needs errors.Wrap with the original error

File: pkg/collect/run.go [hunk 3]
```diff
@@ -137,10 +149,79 @@ func runPod(ctx context.Context, client *kubernetes.Clientset, runCollector *tro
...
+				return errors.Errorf("Secret type kubernetes.io/dockerconfigjson accepts only one argument \".dockerconfigjson\"")
+			}
+			//Then, if data is a path to a config file, it is opened
+			configFile, err := ioutil.ReadFile(v)
+			if err != nil {
+				//If data is not a valid path, we assume data is a base64 encoded config.json file
+				//Client only accepts Json formated files as data, so we decode and indent it (indentation is required)
+				parsedConfig, err := base64.StdEncoding.DecodeString(v)
+				if err != nil {
+					return err
```
Comment (divolgin) [inline pkg/collect/run.go]: `errors.Wrap`

File: pkg/collect/run.go [hunk 4]
```diff
@@ -137,10 +149,79 @@ func runPod(ctx context.Context, client *kubernetes.Clientset, runCollector *tro
...
+			}
+		}
+		secret := corev1.Secret{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "Secret",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:         imagePullSecret.Name,
+				GenerateName: "tmpsecret",
```
Comment (divolgin) [inline pkg/collect/run.go]: let's change `tmpsecret` to `troubleshoot`

File: pkg/collect/run.go [hunk 5]
```diff
@@ -137,10 +149,79 @@ func runPod(ctx context.Context, client *kubernetes.Clientset, runCollector *tro
...
+				Namespace:    pod.Namespace,
+			},
+			Data: data,
+			Type: corev1.SecretType(imagePullSecret.SecretType),
+		}
+		created, err := client.CoreV1().Secrets(pod.Namespace).Create(ctx, &secret, metav1.CreateOptions{})
+		if err != nil {
+			return err
```
Comment (divolgin) [inline pkg/collect/run.go]: `errors.Wrap`

**Comment by matias-manavella-nimble** [comment]
@divolgin Good morning Dimitriy! I made the modifications you suggested.

**Comment by laverya** [comment]
Why not just add an entire pod or container spec here, instead of adding each component one by one?

**Comment by matias-manavella-nimble** [comment]
@laverya It could be done, we could even add the possibility to use a link to a yaml file if necessary. 

Should I start working on that?

**Comment by laverya** [comment]
no, I think that would be a seperate PR
let me give this a review

**Pull request merged**